tests/test_overdrive.py

```diff
@@ -1182,48 +1193,174 @@ def update_licensepool(self, book_id):
+        # mock API delivered the first three LicensePools from the
+        # queue.
         eq_([(1, lp1),(2, lp2),(3, lp3)], api.update_licensepool_calls)
 
-        # At that point we gave up because we'd gotten two consecutive
-        # books from Overdrive that contained no new information.
+        # After each book was processed, should_stop was called, using
+        # the LicensePool, the start date, plus information about
+        # whether the LicensePool was changed (or created) during
+        # update_licensepool().
+        eq_(
+            [(start, 1, True),
+             (start, 2, False),
+             (start, 3, True)],
+            monitor.should_stop_calls
+        )
 
-        # The fourth (bogus) LicensePool is still in api.licensepools
+        # should_stop returned True on the third call, and at that
+        # point we gave up.
+
+        # The fourth (bogus) LicensePool is still in api.licensepools,
+        # because we never asked for it.
         eq_([lp4], api.licensepools)
 
-        # TODO: Verify that a DISTRIBUTOR_TITLE_ADD event was gathered
-        # for the newly discovered license pool.
+        # A single analytics event was sent out, for the first LicensePool,
+        # the one that update_licensepool said was new.
+        [[library, licensepool, event, last_checked]] = monitor.analytics.events
+
+        # The event commemerates the addition of this LicensePool to the
+        # collection.
+        eq_(lp1.collection.libraries, [library])
+        eq_(lp1, licensepool)
+        eq_(CirculationEvent.DISTRIBUTOR_TITLE_ADD, event)
+        eq_(lp1.last_checked, last_checked)
 
         # The incoming TimestampData object was updated with
         # a summary of what happened.
-        eq_("Books processed: 3.", progress.achievements)
+        #
+        # We processed four books: 1, 2, None (which was ignored)
+        # and 3.
+        eq_("Books processed: 4.", progress.achievements)
+
+
+class TestNewTitlesOverdriveCollectionMonitor(OverdriveAPITest):
+
+    def test_recently_changed_ids(self):
+        class MockAPI(object):
+            def __init__(self, *args, **kwargs):
+                pass
+            def all_ids(self):
+                return "all of the ids"
+
+        monitor = NewTitlesOverdriveCollectionMonitor(
+            self._db, self.collection, api_class=MockAPI
+        )
+        eq_("all of the ids", monitor.recently_changed_ids(object(), object()))
+
+    def test_should_stop(self):
+        monitor = NewTitlesOverdriveCollectionMonitor(
+            self._db, self.collection, api_class=MockOverdriveAPI
+        )
+
+        m = monitor.should_stop
+
+        # If the monitor has never run before, we need to keep going
+        # until we run out of books.
+        eq_(False, m(None, object(), object()))
+        eq_(False, m(monitor.NEVER, object(), object()))
+
+        # If information is missing or invalid, we assume that we
```

Comment:
Is this because the missing or invalid data is what is being updated? Or just safe to keep going because there are more books to process?

Reply:
More like "there are more books to process". This shouldn't happen, but if Overdrive provides bad information for a book, I want this script to keep going until it either gets good information, or it reaches the end of the list (which will make it act like the old script -- inefficient but complete).